src/rust/collections/id_map.rs

```diff
@@ -0,0 +1,154 @@
+    /// succeeds, so DO NOT use this function to check for the existance of a particular key. We expect the user to use
+    /// nother data structure for validity.
+    pub fn get(&self, external_id: &E) -> Option<I> {
+        // If we are not obfuscating ids, just return the external id.
+        if DIRECT_MAPPING {
+            Some(<E as Into<u64>>::into(*external_id).into())
+        } else {
+            self.ids.get(external_id).copied()
+        }
+    }
+
+    #[allow(dead_code)]
+    /// Insert a mapping between a specified external and internal id. If we are using a direct mapping,
+    /// then this is a no op.
+    pub fn insert(&mut self, external_id: E, internal_id: I) -> Option<I> {
+        if DIRECT_MAPPING {
+            None
+        } else {
+            self.ids.insert(external_id, internal_id)
+        }
+    }
+
+    /// Remove a mapping between a specificed external and internal id. If the mapping exists, then return the internal
+    /// id mapped to the external id. If we are using a direct mapping, then this is a no op.
+    pub fn remove(&mut self, external_id: &E) -> Option<I> {
+        if DIRECT_MAPPING {
+            Some(<E as Into<u64>>::into(*external_id).into())
+        } else {
+            self.ids.remove(external_id)
+        }
+    }
+
+    /// Generate a new id and insert the mapping to the internal id. If the id is currently in use, keep generating
+    /// until we find an unused id (up to a maximum number of tries). If we are using a direct mapping, then just
+    /// return the internal id without generating a new id or inserting a mapping.
+    pub fn insert_with_new_id(&mut self, internal_id: I) -> E {
+        // If we are not obfuscating ids, just return the external id.
+        if DIRECT_MAPPING {
+            return E::from(internal_id.into());
+        }
+
+        if RANDOMIZE {
+            // Otherwise, allocate a new external id.
+            for _ in 0..MAX_RETRIES_ID_ALLOC {
+                let external_id: E = E::from(self.rng.next_u64());
+                if !self.ids.contains_key(&external_id) {
+                    self.ids.insert(external_id, internal_id);
+                    return external_id;
+                }
+            }
+            panic!("Could not find a valid task id");
+        } else {
+            // Use a Fibonacci sequence.
+            let id: u64 = self.current_id;
+            // Roll around.
+            self.current_id = if self.current_id < u64::MAX - self.last_id {
+                self.current_id + self.last_id
+            } else {
+                self.last_id - (u64::MAX - self.current_id)
+            };
+            self.last_id = id;
+            let external_id: E = E::from(id);
+            if self.ids.insert(external_id, internal_id).is_some() {
+                panic!("Should not have a previous task with this id");
+            }
+            external_id
```

Comment:
What are the motivations for using a Fibonnacci sequence in contrast to a simple arithmethic progression with step 1 starting at 0?